torch/csrc/autograd/python_saved_variable_hooks.cpp

```diff
@@ -46,16 +46,22 @@ namespace torch { namespace autograd {
   }
 
   // NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
-  PyObject* PyDefaultSavedVariableHooks::pack_hook_ = nullptr;
+  std::atomic<PyObject*> PyDefaultSavedVariableHooks::pack_hook_(nullptr);
   // NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
-  PyObject* PyDefaultSavedVariableHooks::unpack_hook_ = nullptr;
+  std::atomic<PyObject*> PyDefaultSavedVariableHooks::unpack_hook_(nullptr);
 
   void PyDefaultSavedVariableHooks::set_hooks(py::function &pack_hook, py::function &unpack_hook) {
+    PyObject* expected = nullptr;
     TORCH_CHECK(!pack_hook_ && !unpack_hook_,
         "Setting default hooks but they have already been set. "
         "Hint: only one pair of hooks is allowed at a time.");
-    pack_hook_ = pack_hook.release().ptr();
-    unpack_hook_ = unpack_hook.release().ptr();
+    PyObject* pack_hook_ptr = pack_hook.release().ptr();
+    PyObject* unpack_hook_ptr = unpack_hook.release().ptr();
+    if (!pack_hook_.compare_exchange_strong(expected, pack_hook_ptr) ||
+        !unpack_hook_.compare_exchange_strong(expected, unpack_hook_ptr)) {
+      reset_hooks();
+      TORCH_INTERNAL_ASSERT(false); // race condition
```

Comment:
This should contain an error message so that the user is getting some context into what's happening